File: src/main/java/co/herod/adbwrapper/BusSubject.java
package co.herod.adbwrapper;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;
import org.jetbrains.annotations.NotNull;

public abstract class BusSubject<T> extends Subject<T> implements Observer<T> {

    private final PublishSubject<T> processSubject = PublishSubject.create();

    @Override
    protected void subscribeActual(@NotNull final Observer<? super T> observer) {
        processSubject.subscribeActual(observer);
    }

    @Override
    public boolean hasObservers() {
        return processSubject.hasObservers();
    }

    @Override
    public boolean hasThrowable() {
        return processSubject.hasThrowable();
    }

    @Override
    public boolean hasComplete() {
        return processSubject.hasComplete();
    }

    @Override
    public Throwable getThrowable() {
        return processSubject.getThrowable();
    }

    @Override
    public void onSubscribe(final Disposable d) {
        processSubject.onSubscribe(d);
    }

    @Override
    public void onNext(final T s) {
        processSubject.onNext(s);
    }

    @Override
    public void onError(final Throwable e) {
        processSubject.onError(e);
    }

    @Override
    public void onComplete() {
        // ignored
    }
}

File: adb-kotlin/src/main/java/co/herod/adbwrapper/BusSubject.kt
package co.herod.adbwrapper

import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject


abstract class BusSubject<T> : Subject<T>(), Observer<T> {

    private val processSubject = PublishSubject.create<T>()

    public override fun subscribeActual(observer: Observer<in T>) {
        processSubject.subscribeActual(observer)
    }

    override fun hasObservers(): Boolean {
        return processSubject.hasObservers()
    }

    override fun hasThrowable(): Boolean {
        return processSubject.hasThrowable()
    }

    override fun hasComplete(): Boolean {
        return processSubject.hasComplete()
    }

    override fun getThrowable(): Throwable? {
        return processSubject.throwable
    }

    override fun onSubscribe(d: Disposable) {
        processSubject.onSubscribe(d)
    }

    override fun onNext(s: T) {
        processSubject.onNext(s)
    }

    override fun onError(e: Throwable) {
        processSubject.onError(e)
    }

    override fun onComplete() {
        // ignored
    }
}
